File: ios/localasset-core/Sources/LocalAssetCore/registry/InMemoryHandleRegistry.swift
import Foundation

public class InMemoryHandleRegistry: HandleRegistry {
    private struct StoredHandle {
        let record: ResourceHandleRecord
        let createdAtMillis: Int64
    }

    private var items: [String: StoredHandle] = [:]
    private let lock = NSLock()
    private let host: String
    private let pathPrefix: String
    private let nowMillis: () -> Int64

    public init(
        host: String = defaultHandleHost,
        pathPrefix: String = "/handles",
        nowMillis: @escaping () -> Int64 = currentMillis
    ) {
        self.host = host
        self.pathPrefix = pathPrefix
        self.nowMillis = nowMillis
    }

    public func register(handle: ResourceHandle) -> String {
        let safe = handle.fileName.replacingOccurrences(
            of: "[^A-Za-z0-9._-]", with: "_", options: .regularExpression
        )
        let uri = buildUri(token: UUID().uuidString, fileName: safe)
        let size: Int64? = {
            if case .bytes(let data) = handle.source { return Int64(data.count) }
            return nil
        }()
        let record = ResourceHandleRecord(
            resourceUri: uri,
            source: handle.source,
            fileName: handle.fileName,
            mimeType: handle.mimeType,
            size: size,
            metadata: handle.metadata,
            ttlMillis: handle.ttlMillis
        )
        lock.withLock { items[uri] = StoredHandle(record: record, createdAtMillis: nowMillis()) }
        return uri
    }

    public func resolve(resourceUri: String) throws -> ResourceHandleRecord {
        try lock.withLock {
            guard let stored = items[resourceUri] else {
                throw ResourceException(.resolutionError, "Unknown or expired resourceUri: \(resourceUri)")
            }
            if let ttl = stored.record.ttlMillis,
               nowMillis() - stored.createdAtMillis >= ttl {
                items.removeValue(forKey: resourceUri)
                throw ResourceException(.resolutionError, "Unknown or expired resourceUri: \(resourceUri)")
            }
            return stored.record
        }
    }

    public func remove(resourceUri: String) {
        lock.withLock { _ = items.removeValue(forKey: resourceUri) }
    }

    public func cleanup(nowMillis: Int64) {
        lock.withLock {
            items = items.filter { _, stored in
                stored.record.ttlMillis.map { stored.createdAtMillis + $0 > nowMillis } ?? true
            }
        }
    }

    private func buildUri(token: String, fileName: String) -> String {
        let prefix = pathPrefix.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let p = prefix.isEmpty ? "handles" : prefix
        return "local-asset://\(host)/\(p)/\(token)/\(fileName)"
    }
}

File: android/localasset-core/src/main/kotlin/io/github/xesam/android/localasset/core/registry/InMemoryHandleRegistry.kt
package io.github.xesam.android.localasset.core.registry

import io.github.xesam.android.localasset.core.api.HandleRegistry
import io.github.xesam.android.localasset.core.error.ResourceErrorCategory
import io.github.xesam.android.localasset.core.error.ResourceException
import io.github.xesam.android.localasset.core.model.ResourceHandle
import io.github.xesam.android.localasset.core.model.ResourceHandleRecord
import io.github.xesam.android.localasset.core.model.ResourceSource
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

class InMemoryHandleRegistry(
    private val host: String = HandleRegistry.DEFAULT_HANDLE_HOST,
    private val pathPrefix: String = "/handles",
    private val nowMillis: () -> Long = { System.currentTimeMillis() },
) : HandleRegistry {
    private val items = ConcurrentHashMap<String, StoredHandle>()

    override fun register(handle: ResourceHandle): String {
        val safeFileName = handle.fileName.replace(Regex("[^A-Za-z0-9._-]"), "_")
        val resourceUri = buildUri(
            token = UUID.randomUUID().toString(),
            fileName = safeFileName,
        )
        // size is only known for Bytes handles; for FilePath it is probed at resolve time, not here.
        val size: Long? = (handle.source as? ResourceSource.Bytes)?.value?.size?.toLong()
        items[resourceUri] = StoredHandle(
            record = ResourceHandleRecord(
                resourceUri = resourceUri,
                source = handle.source,
                fileName = handle.fileName,
                mimeType = handle.mimeType,
                size = size,
                metadata = handle.metadata,
                ttlMillis = handle.ttlMillis,
            ),
            createdAtMillis = nowMillis(),
        )
        return resourceUri
    }

    override fun resolve(resourceUri: String): ResourceHandleRecord {
        return items.compute(resourceUri) { _, stored ->
            if (stored == null) return@compute null
            val ttlMillis = stored.record.ttlMillis
            if (ttlMillis != null && nowMillis() - stored.createdAtMillis >= ttlMillis) {
                null
            } else {
                stored
            }
        }?.record ?: throw ResourceException(
            category = ResourceErrorCategory.RESOLUTION_ERROR,
            message = "Unknown or expired resourceUri: $resourceUri",
        )
    }


    override fun cleanup(nowMillis: Long) {
        items.entries.removeIf { (_, stored) ->
            stored.record.ttlMillis?.let { stored.createdAtMillis + it <= nowMillis } == true
        }
    }

    override fun remove(resourceUri: String) {
        items.remove(resourceUri)
    }

    private fun buildUri(token: String, fileName: String): String {
        val normalizedPrefix = pathPrefix.trim('/').takeIf { it.isNotBlank() } ?: "handles"
        return "local-asset://$host/$normalizedPrefix/$token/$fileName"
    }

    private data class StoredHandle(
        val record: ResourceHandleRecord,
        val createdAtMillis: Long,
    )
}
